sky/data/csync/redirect-fuse.c

```diff
@@ -0,0 +1,542 @@
+/*
+  FUSE: Filesystem in Userspace
+  Copyright (C) 2001-2007  Miklos Szeredi <miklos@szeredi.hu>
+  Copyright (C) 2011       Sebastian Pipping <sebastian@pipping.org>
+
+  This program can be distributed under the terms of the GNU GPLv2.
+  See the file COPYING.
+*/
+
+/** @file
+ *
+ * This file system redirects the write and read on mountpoint to other
+ * designated directories to work along with sync daemon and cloud storage
+ * mount under sky/data/csync/sky_csync.py. The code is inspired by an example
+ * provided at:
+ * 		https://github.com/libfuse/libfuse/blob/master/example/passthrough.c		
+ * 
+ * Compile with
+ *
+ *     gcc -Wall -D_FILE_OFFSET_BITS=64 redirect-fuse.c set.c `pkg-config fuse3 --cflags --libs` -o redirect-fuse
+ *
+ * ## Source code ##
+ * \include redirect-fuse.c
+ */
+
+
+#define FUSE_USE_VERSION 31
+
+#define _GNU_SOURCE
+
+#ifdef linux
+#endif
+
+#include <fuse.h>
+#include <libgen.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <dirent.h>
+#include <errno.h>
+#include <stdlib.h>
+#ifdef __FreeBSD__
+#endif
+#include <sys/time.h>
+
+#include "set.h"
+
+static int fill_dir_plus = 0;
+static char *mountpoint_path = NULL;
+static char *csync_read_path = NULL;
+static char *csync_write_path = NULL;
+
+
+char* join_path(const char* base_path, const char* partial_path) {
+    if (base_path == NULL || partial_path == NULL) {
+        return NULL;
+    }
+
+    // Allocate enough space for the full path
+	// +2 for '/' and '\0'
+    char* full_path = malloc(strlen(base_path) + strlen(partial_path) + 2);
+    if (full_path == NULL) {
+        return NULL; // Allocation failed
+    }
+
+    strcpy(full_path, base_path);
+    strcat(full_path, "/");
+    strcat(full_path, partial_path);
+
+    return full_path;
+}
+
+int file_exists(const char* path) {
+    struct stat buffer;
+    return (stat(path, &buffer) == 0); 
+}
+
+char* full_path(const char* partial_path, int write, int read) {
+    // Check and remove leading '/' from partial_path if present
+    if (partial_path[0] == '/') {
+        partial_path++;  // Move the pointer to skip the leading '/'
+    }
+
+    char* full_write_path = join_path(csync_write_path, partial_path);
+    char* full_read_path = join_path(csync_read_path, partial_path);
+    char* result = NULL;
+
+	if (write) {
+        result = full_write_path;
+        free(full_read_path);
+	} else if (read) {
+        result = full_read_path;
+        free(full_write_path);
+	} else if (file_exists(full_write_path)) {
+        result = full_write_path;
+        free(full_read_path);
+    } else {
+        result = full_read_path;
+        free(full_write_path);
+    } 
+
+    return result;
+}
```

Comment:
Should we only call join_path if needed? seems like in all branches we are allocation 2 and free 1. why not only allocate once